src/server/server.ts

```diff
@@ -433,23 +433,25 @@ namespace ts.server {
                 return;
             }
 
-            if (this.activeRequestCount > 0) {
-                this.activeRequestCount--;
-            }
-            else {
-                Debug.fail("Received too many responses");
-            }
-
-            while (this.requestQueue.length > 0) {
-                const queuedRequest = this.requestQueue.shift();
-                if (this.requestMap.get(queuedRequest.operationId) === queuedRequest) {
-                    this.requestMap.delete(queuedRequest.operationId);
-                    this.scheduleRequest(queuedRequest);
-                    break;
+            if (response.kind === ActionSet) {
+                if (this.activeRequestCount > 0) {
+                    this.activeRequestCount--;
+                }
+                else {
+                    Debug.fail("Received too many responses");
                 }
 
-                if (this.logger.hasLevel(LogLevel.verbose)) {
-                    this.logger.info(`Skipping defunct request for: ${queuedRequest.operationId}`);
+                while (this.requestQueue.length > 0) {
+                    const queuedRequest = this.requestQueue.shift();
+                    if (this.requestMap.get(queuedRequest.operationId) === queuedRequest) {
+                        this.requestMap.delete(queuedRequest.operationId);
+                        this.scheduleRequest(queuedRequest);
+                        break;
+                    }
+
+                    if (this.logger.hasLevel(LogLevel.verbose)) {
+                        this.logger.info(`Skipping defunct request for: ${queuedRequest.operationId}`);

```

Comment:
This still seems like a hazard. You're effectively decrementing by one above to keep track of request/response pairings, yet in this loop you just shift off requests until you find the one you want (with a log "info" message that a request was skipped). Wouldn't this get the activeRequestCount out of sync if it skips them?

Reply:
Oh, don't mind me. I see this is scheduling the next one from the queue, so should be good.

Reply:
Though one concern/hazard: This while loop previously was always entered to schedule the next request, other than for the 3 event types explicitly handled in the "if" blocks above. Now you only enter this loop to continue scheduling work "if (response.kind === ActionSet)". Shouldn't it continue to always enter this loop except for any events? (Even if only for future proofing?)

Reply:
I can change the check to `!invalidate` if you think that would be more future proof.  Otherwise, it's one-in-one-out, so it doesn't make sense to pull anything off the queue if the message doesn't indicate the completion of a request.

Reply:
Seems like letting the type system verify exhaustive checks on the string literal types in "TypingInstallerResponse" would be safest to ensure any additions/changes to "response.kind" aren't missed (e.g. https://stackoverflow.com/questions/39419170/how-do-i-check-that-a-switch-block-is-exhaustive-in-typescript ). Possible here?